tket/src/passes/squash_borrow.rs

```diff
@@ -0,0 +1,708 @@
+//! Reorder and squash pairs of borrow and return nodes where possible.
+
+use derive_more::{Display, Error};
+use hugr::algorithms::ComposablePass;
+use hugr::extension::prelude::ConstUsize;
+use hugr::extension::simple_op::MakeExtensionOp;
+use hugr::hugr::hugrmut::HugrMut;
+use hugr::ops::{OpTag, OpTrait, OpType, Value};
+use hugr::std_extensions::arithmetic::conversions::ConvertOpDef;
+use hugr::std_extensions::arithmetic::int_types::ConstInt;
+use hugr::std_extensions::collections::borrow_array::{BArrayUnsafeOpDef, BorrowArray};
+use hugr::types::{Term, Type};
+use hugr::{HugrView, IncomingPort, Node, OutgoingPort, PortIndex, Wire};
+use itertools::{Either, Itertools};
+
+use std::collections::hash_map::Entry;
+use std::collections::{HashMap, HashSet, VecDeque};
+
+/// Identifies array types and borrows/returns on them
+pub trait IsBorrowReturn: Clone {
+    /// Determine if the given node is a borrow or return node, and if so,
+    /// return the ports identifying the operands. May also return
+    /// * [BRAction::Clobber] if the op is neither a borrow nor return but
+    ///   affects borrowedness of only some indices.
+    /// * `Ok(None)` if the node is not a borrow/return, and we should not
+    ///   elide any borrow/return pairs across this node.
+    ///
+    /// # Errors
+    ///
+    /// [BorrowAnalysisError] if the analysis cannot safely determine whether the
+    /// node is a borrow/return with known operands. (Generally, returning `Ok(None)`
+    /// handles most situations by preventing optimization across a particular node;
+    /// however, returning an error prevents optimization across the wider DSG.)
+    fn is_borrow_return<H: HugrView>(
+        &self,
+        node: H::Node,
+        hugr: &H,
+    ) -> Result<Option<BorrowReturnPorts>, BorrowAnalysisError>;
+
+    /// If the specified type is an array that can be borrowed from,
+    /// return `Some` of:
+    ///   * `Some(size)` if the array's size is statically known.
+    ///   * `None` if the array's size is not statically known.
+    ///
+    /// Otherwise (if the type is not such an array), return `None`.
+    fn get_array_size(&self, ty: &Type) -> Option<Option<u64>>;
```

Comment:
The nested option is a bit annoying for readability, could be worth a dedicated enum